doc/user_guide/model.rst

```diff
@@ -659,65 +659,66 @@ Fitting the model to the data
-    +--------------------------+--------+------------------+----------------------+--------+
-    | "Powell"                 |  No    | No               | 'ls', 'ml', 'custom' | local  |
-    +--------------------------+--------+------------------+----------------------+--------+
-    | "CG"                     |  No    | No               | 'ls', 'ml', 'custom' | local  |
-    +--------------------------+--------+------------------+----------------------+--------+
-    | "BFGS"                   |  No    | No               | 'ls', 'ml', 'custom' | local  |
-    +--------------------------+--------+------------------+----------------------+--------+
-    | "Newton-CG"              |  No    | No               | 'ls', 'ml', 'custom' | local  |
-    +--------------------------+--------+------------------+----------------------+--------+
-    | "L-BFGS-B"               |  Yes   | No               | 'ls', 'ml', 'custom' | local  |
-    +--------------------------+--------+------------------+----------------------+--------+
-    | "TNC"                    |  Yes   | No               | 'ls', 'ml', 'custom' | local  |
-    +--------------------------+--------+------------------+----------------------+--------+
-    | "Differential Evolution" |  Yes   | No               | 'ls', 'ml', 'custom' | global |
-    +--------------------------+--------+------------------+----------------------+--------+
-
+.. table:: Features of supported curve-fitting optimizers.
+
+    +------------------------------+--------+-----------+--------+----------------+--------+
+    | Optimizer                    | Bounds | Gradients | Errors | Method         | Type   |
+    +==============================+========+===========+========+================+========+
+    | ``"leastsq"``                |  Yes   | Yes       | Yes    | Only ``"ls"``  | local  |
+    +------------------------------+--------+-----------+--------+----------------+--------+
+    | ``"mpfit"`` *                |  Yes   | Yes       | Yes    | Only ``"ls"``  | local  |
+    +------------------------------+--------+-----------+--------+----------------+--------+
+    | ``"odr"``                    |  No    | Yes       | Yes    | Only ``"ls"``  | local  |
+    +------------------------------+--------+-----------+--------+----------------+--------+
+    | ``"Nelder-Mead"``            |  No    | No        | No     | All            | local  |
+    +------------------------------+--------+-----------+--------+----------------+--------+
+    | ``"L-BFGS-B"``               |  Yes   | Yes       | Yes    | All            | local  |
+    +------------------------------+--------+-----------+--------+----------------+--------+
+    | ``"Differential Evolution"`` |  Yes   | No        | No     | All            | global |
+    +------------------------------+--------+-----------+--------+----------------+--------+
+
+\* ``mpfit`` is deprecated since HyperSpy 1.6. Use ``"leastsq"`` instead.
```

Comment:
Hmmm, that'll probably be because `mpfit` uses the Levenberg–Marquardt algorithm with its own implementation of bounds, while `leastsq` with bounds actually uses the Trust Region Reflective algorithm (`method="trf"`).

There are a couple of options here, both of which could preserve `mpfit`.

1. Deprecate the options `["leastsq", "mpfit"]` with e.g. `["lm", "trf", "dogbox"]` with clearer documentation.
2. Don't do anything, leave `"mpfit"` as it is.

Option (1) is nice because the default fitter for `m.fit()` could then be `None`, and we could smartly suggest an appropriate method depending on the type of function, bounds, and gradients.

Option (2) is OK, but I think it would be better to deprecate `mpfit` for `lm` (Levenberg–Marquardt).

Reply:
Since mpfit is vendored and it is not supported since a while, at some point, it may break. If we don't have a good reason to keep mpfit, we should deprecate it.

Reply:
With a little bit of work it can probably support passing a custom minimizer (like we did for `s.decomposition()` and sklearn objects), so you could ultimately use `lmfit`, `scikit-optimize` or `openopt/copt` if desired. That could be option 3.

Reply:
That sounds excellent, but maybe something for a different PR? A possible problem (for which solutions exist) with that approach is that it won't work for parallel distributed fits since the library will need to be present in all the nodes.

Regarding mpfit, as mentioned above, in the past, I have obtained better results with it than with bounded ``scipy.optimize.least_squares``, so, unless we show that this not actually generally the case, I suggest that we keep it unless it gets broken beyond easy repair at some point. it seems that a maintained version of the code can be found in https://github.com/segasai/astrolibpy . However, it is not distributed in pypi as far as I can tell. Ours is also a maintained version of the code and so-far it has been easy to fix it every time that changes in numpy broke something. For the future (i.e. after The Split) If we want to continue offering mpfit, I suggest to contact @segasai to check if he would be happy to publish his code (that is the original code from which we forked our version a long ago) in pypi and conda-forge.

I like @tjof2's 1. option. Currently it seems that even for unbounded optimization the default bounding method is "trf", while it would be more efficient to use "lm". An ``auto`` mode could fix this, while leaving the user the option to choose a different method if required. If we keep mpfit, we could add ``mpfit`` to the list of bounds methods. Also, the optimizer in that case could be "LM" instead of "leastsq" as it is now, since "leastsq" is rather the loss-funtion and we could use all other optimizers to solve the same problem.

Reply:
Thanks @francisco-dlp  - I also like option 1. I'll work on that.

Reply:
OK, maybe it would be worth trying to sort out the distribution of mpfit or even better integrate it in scipy!

Reply:
That's a good point: if the presumed advantage comes from a particular way of implementing the bounds, maybe that can be added to scipy's ``least_squares`` as an extra bounds method.

Reply:
I would imagine that scipy probably won't want it :)

Reply:
I've found a nicer way to keep mpfit rather than deprecating (for now)

Reply:
> Since mpfit is vendored and it is not supported since a while, at some point, it may break. If we don't have a good reason to keep mpfit, we should deprecate it

The code is actually much simpler than it looks @ericpre. It probably looks bad due to all the old comments, old Fortran comments and old Python style (or lack of...). I've actually tidied it up a lot, I can easily push it in a separate PR. Since it's written in pure Python (unlike `scipy.optimize.least_squares` which wraps Fortran), it can be a little slower when optimizing, but the bound constraints are useful, so keeping it (for now at least) makes sense.